In [1]:
#importaciones de datos
import pandas as pd

#importación de src codigo de soporte con las funciones necesarias para la limpieza
import sys
sys.path.append("..")
from src import sp_eda_limpieza as sp

In [2]:
#ajustes de visualización
pd.set_option('display.max_columns', None)

In [3]:
#lectura de los ficheros de datos
df_bankadditional_raw=pd.read_csv(r'C:\Users\matis\Desktop\EDA_MARKETING\data\fuentes de datos\bank-additional.csv')

In [59]:
sheets=pd.read_excel(r'C:\Users\matis\Desktop\EDA_MARKETING\data\fuentes de datos\customer-details.xlsx',sheet_name=None)
df_customerdetails_raw=pd.concat(sheets.values(),ignore_index=True)

In [60]:
#se cran las variables que almacenan la copia del fichero original. estas se manipularán para ser limpiadas
df_bankadditional=df_bankadditional_raw.copy()
df_customerdetails=df_customerdetails_raw.copy()

In [61]:
#Eda preeliminar del fichero customer details
sp.eda_preliminar(df_customerdetails)

Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
18441  108057        2         1  2012-10-09                  7   
30016   20759        1         1  2014-01-03                 22   
10680  108224        0         2  2012-06-03                 11   
36018  108272        0         2  2014-12-07                 32   
39008   47511        1         0  2014-12-03                 10   

                                         ID  
18441  8a81370d-ff25-4b2d-80f9-6e497a9ff178  
30016  4afc7d59-aff1-4223-9286-7cfaaccababd  
10680  7a14695c-f993-411a-b2bc-9a13366498fd  
36018  60081fae-4ff4-4e46-afe0-c5a3e03ebfcc  
39008  d5e3114b-3a3a-4f28-bec9-f2fc628209a6

--------------------
INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Income             43170 non-null  int64         
 1   Kidhome            43170 non-null  int64         
 2   Teenhome           43170 non-null  int64         
 3   Dt_Customer        43170 non-null  datetime64[ns]
 4   NumWebVisitsMonth  43170 non-null  int64         
 5   ID                 43170 non-null  object        
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 2.0+ MB


None

--------------------
NULOS


Income               0.0
Kidhome              0.0
Teenhome             0.0
Dt_Customer          0.0
NumWebVisitsMonth    0.0
ID                   0.0
dtype: float64

--------------------
duplicados
0
VALUE COUNT


ID
9727bc06-c11a-461a-a5bb-3d210467cc2a    1
089b39d8-e4d0-461b-87d4-814d71e0e079    1
5006b75e-18bb-46e9-ae1a-89f7eb79cc8a    1
e554d2ad-6c83-4603-93d4-fc596b807eff    1
3959015a-9713-4877-9a43-740e34c29eaf    1
                                       ..
5e3483e5-236d-437d-8351-541f9d09b9dd    1
d63ede72-0b6d-45b1-8872-385ac6897f65    1
eca60b76-70b6-4077-80ba-bc52e8ebb0eb    1
9991fafb-4447-451a-8be2-b0df6098d13e    1
3f9f49b5-e410-4948-bf6e-f9244f04918b    1
Name: count, Length: 43170, dtype: int64

--------------------
VALORES NUMÉRICOS


count                           mean  \
Income             43170.0                   93227.389622   
Kidhome            43170.0                       1.004749   
Teenhome           43170.0                       0.998633   
Dt_Customer          43170  2013-05-11 22:09:47.407922176   
NumWebVisitsMonth  43170.0                       16.59129   

                                   min                  25%  \
Income                          5841.0              49608.0   
Kidhome                            0.0                  0.0   
Teenhome                           0.0                  0.0   
Dt_Customer        2012-01-01 00:00:00  2012-07-16 00:00:00   
NumWebVisitsMonth                  1.0                  9.0   

                                   50%                  75%  \
Income                         93009.5             136740.5   
Kidhome                            1.0                  2.0   
Teenhome                           1.0                  2.0   
Dt_Customer        2013-03-01 00:00:00  2014-03-28 00:00:00   
NumWebVisitsMonth                 17.0                 25.0   

                                   max           std  
Income                        180802.0  50498.181989  
Kidhome                            2.0      0.815996  
Teenhome                           2.0      0.816008  
Dt_Customer        2014-12-31 00:00:00           NaN  
NumWebVisitsMonth                 32.0      9.241769

## Análisis
- Borrar columna unnamed:0
- nulos 0
- duplicados 0
- categorías: todo ok
## To do
- Borrar columa unnamed:0

In [62]:
sp.borrar_unnamed(df_customerdetails)

Columnas originales: Index(['Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'NumWebVisitsMonth',
       'ID'],
      dtype='object')
La columna 'Unnamed: 0' no se encuentra en el DataFrame


Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  \
25143  170312        0         1  2013-03-15                  6   

                                         ID  
25143  1b070063-621d-4093-8275-21f6b30890f8

In [63]:
#Eda preeliminar del fichero bankadditional
sp.eda_preliminar(df_bankadditional)

Unnamed: 0   age           job   marital            education  default  \
40899       41066   NaN    technician   MARRIED    university.degree      0.0   
21689       21786   NaN        admin.    SINGLE    university.degree      0.0   
19046       19143  44.0    technician  DIVORCED    university.degree      NaN   
19248       19345  51.0  entrepreneur   MARRIED             basic.9y      0.0   
22783       22880  32.0    technician   MARRIED  professional.course      0.0   

       housing  loan   contact  duration  campaign  pdays  previous  \
40899      0.0   0.0  cellular       119         1    999         3   
21689      0.0   0.0  cellular       162         1    999         0   
19046      0.0   0.0  cellular       117         1    999         0   
19248      0.0   0.0  cellular       135         1    999         0   
22783      1.0   1.0  cellular        69         2    999         0   

          poutcome  emp.var.rate cons.price.idx cons.conf.idx euribor3m  \
40899      FAILURE          -1.1         94,767         -50,8      1,05   
21689  NONEXISTENT           1.4         93,444         -36,1     4,964   
19046  NONEXISTENT           1.4         93,444         -36,1     4,968   
19248  NONEXISTENT           1.4         93,444         -36,1     4,967   
22783  NONEXISTENT           1.4         93,444         -36,1     4,965   

      nr.employed   y                date  latitude  longitude  \
40899      4963,6  no   19-diciembre-2017    32.407    -94.287   
21689      5228,1  no  27-septiembre-2018    44.069   -102.012   
19046      5228,1  no       29-abril-2016    28.014   -122.098   
19248      5228,1  no       28-julio-2018    32.375    -88.002   
22783      5228,1  no        6-julio-2019    42.899   -105.806   

                                        id_  
40899  b189da74-1967-40f5-abfd-abb719aa8bd6  
21689  acaec973-9a13-402b-8a28-d77da7fca5bc  
19046  fc8766a0-91fd-48f8-bbf4-d2d8e17b96b4  
19248  e688a12b-f301-4131-a323-a47a559d082a  
22783  a89f8310-1ec6-416c-ab7e-f996c3bb2bd1

--------------------
INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  object 
 3   marital         42915 non-null  object 
 4   education       41193 non-null  object 
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  object 
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  int64  
 12  previous        43000 non-null  int64  
 13  poutcome        43000 non-null  object 
 14  emp.var.rate    43000 non-null  float64
 15  cons.price.idx  42529 non-null  object 
 16  cons.conf.idx   43000 non-null  object 
 17  eurib

None

--------------------
NULOS


Unnamed: 0         0.00
age               11.91
job                0.80
marital            0.20
education          4.20
default           20.89
housing            2.39
loan               2.39
contact            0.00
duration           0.00
campaign           0.00
pdays              0.00
previous           0.00
poutcome           0.00
emp.var.rate       0.00
cons.price.idx     1.10
cons.conf.idx      0.00
euribor3m         21.53
nr.employed        0.00
y                  0.00
date               0.58
latitude           0.00
longitude          0.00
id_                0.00
dtype: float64

--------------------
duplicados
0
VALUE COUNT


job
admin.           10873
blue-collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self-employed     1489
housemaid         1123
unemployed        1063
student            903
Name: count, dtype: int64

--------------------


marital
MARRIED     25999
SINGLE      12105
DIVORCED     4811
Name: count, dtype: int64

--------------------


education
university.degree      12722
high.school             9925
basic.9y                6309
professional.course     5477
basic.4y                4356
basic.6y                2386
illiterate                18
Name: count, dtype: int64

--------------------


contact
cellular     27396
telephone    15604
Name: count, dtype: int64

--------------------


poutcome
NONEXISTENT    37103
FAILURE         4461
SUCCESS         1436
Name: count, dtype: int64

--------------------


cons.price.idx
93,994    7938
93,918    6937
92,893    5985
93,444    5349
94,465    4522
93,2      3731
93,075    2552
92,201     803
92,963     742
92,431     471
92,649     375
94,215     320
94,199     316
92,843     298
92,379     274
93,369     271
94,027     242
94,055     236
93,876     224
94,601     212
92,469     184
92,713     179
93,749     179
94,767     130
93,798      48
92,756      11
Name: count, dtype: int64

--------------------


cons.conf.idx
-36,4    8020
-42,7    7004
-46,2    6057
-36,1    5408
-41,8    4568
-42      3782
-47,1    2581
-31,4     815
-40,8     748
-26,9     477
-30,1     379
-40,3     328
-37,5     318
-50       299
-29,8     277
-34,8     275
-38,3     245
-39,8     239
-40       225
-49,5     214
-33,6     185
-34,6     183
-33       181
-50,8     133
-40,4      48
-45,9      11
Name: count, dtype: int64

--------------------


euribor3m
4,857    2287
4,962    2124
4,963    2019
4,961    1594
4,964     964
         ... 
5,045       1
4,921       1
0,895       1
0,956       1
0,953       1
Name: count, Length: 309, dtype: int64

--------------------


nr.employed
5228,1    16980
5099,1     8937
5191       8020
5195,8     3830
5076,2     1748
5017,5     1133
4991,6      812
5008,7      683
4963,6      665
5023,5      181
5176,3       11
Name: count, dtype: int64

--------------------


y
no     38156
yes     4844
Name: count, dtype: int64

--------------------


date
28-febrero-2018      95
28-febrero-2016      94
28-febrero-2015      92
28-febrero-2017      89
28-febrero-2019      85
                     ..
17-diciembre-2018    11
21-diciembre-2019    10
14-abril-2015        10
4-abril-2017         10
20-noviembre-2015    10
Name: count, Length: 1825, dtype: int64

--------------------


id_
993bbbd6-4dbc-4a40-a408-f91f8462bee6    1
089b39d8-e4d0-461b-87d4-814d71e0e079    1
e9d37224-cb6f-4942-98d7-46672963d097    1
3f9f49b5-e410-4948-bf6e-f9244f04918b    1
9991fafb-4447-451a-8be2-b0df6098d13e    1
                                       ..
66a752e2-b2f1-440c-9a8f-cc3b10d74dd0    1
ca5c0d7a-8bbd-42b3-8351-1074c5be011a    1
8d700df5-0c33-4517-8cf8-92e1c92c9c12    1
ea6b7d04-9271-4c0a-a01f-07795d164aba    1
87b79988-2be5-419d-88f4-56655852c565    1
Name: count, Length: 43000, dtype: int64

--------------------
VALORES NUMÉRICOS


count          mean           std      min          25%  \
Unnamed: 0    43000.0  20651.099047  11868.770913    0.000  10381.75000   
age           37880.0     39.977112     10.437957   17.000     32.00000   
default       34019.0      0.000088      0.009390    0.000      0.00000   
housing       41974.0      0.535998      0.498708    0.000      0.00000   
loan          41974.0      0.155620      0.362499    0.000      0.00000   
duration      43000.0    257.739279    258.666033    0.000    102.00000   
campaign      43000.0      2.567233      2.772294    1.000      1.00000   
pdays         43000.0    962.330953    187.260394    0.000    999.00000   
previous      43000.0      0.174023      0.497366    0.000      0.00000   
emp.var.rate  43000.0      0.077128      1.573898   -3.400     -1.80000   
latitude      43000.0     36.856697      7.225948   24.396     30.61475   
longitude     43000.0    -95.939067     16.752282 -124.997   -110.49425   

                     50%          75%        max  
Unnamed: 0    20642.5000  30930.25000  41187.000  
age              38.0000     47.00000     98.000  
default           0.0000      0.00000      1.000  
housing           1.0000      1.00000      1.000  
loan              0.0000      0.00000      1.000  
duration        179.0000    319.00000   4918.000  
campaign          2.0000      3.00000     56.000  
pdays           999.0000    999.00000    999.000  
previous          0.0000      0.00000      7.000  
emp.var.rate      1.1000      1.40000      1.400  
latitude         36.7610     43.11325     49.384  
longitude       -95.8995    -81.42775    -66.937

## Análisis to do 
- Borrar columna unnamed
- Poner iniciales de las columnas en minúsculas
- conversión de booleanos a sí no: default,housing,loan
- conversión de valores a int: nr.employed
- conversión de valores float: cons.price.idx,cons.conf.ids, euribor3m
- eliminar pdays: el 75% de lso valores son 999
- pasar  a fecha la columna date
- pasar valores a minúsculas: marital, poutcome
- education, cambiar el punto por una _
- cambiar - por _ en job
- cambiar Id_ por ID
- Cambiar Y por Subscribed
- Unir los dos dataframes
- gestión de nulos
- conversión de coordenadas a ciudad, región, país

In [64]:
#borrar la columna unnamed
sp.borrar_unnamed(df_bankadditional)

Columnas originales: Index(['Unnamed: 0', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'latitude',
       'longitude', 'id_'],
      dtype='object')
Columna 'Unnamed: 0' eliminada


age     job  marital          education  default  housing  loan  \
20725  35.0  admin.  MARRIED  university.degree      NaN      0.0   0.0   

        contact  duration  campaign  pdays  previous     poutcome  \
20725  cellular       181         5    999         0  NONEXISTENT   

       emp.var.rate cons.price.idx cons.conf.idx euribor3m nr.employed   y  \
20725           1.4         93,444         -36,1     4,965      5228,1  no   

                 date  latitude  longitude  \
20725  20-agosto-2016    30.625    -97.301   

                                        id_  
20725  728b925b-b2a3-483f-a964-824986652492

In [65]:
#pasar columnas a mayúscula
sp.columnas_mayuscula(df_bankadditional)

Columnas pasadas a mayúscualas


Age          Job  Marital Education  Default  Housing  Loan   Contact  \
25903  36.0  blue-collar  MARRIED  basic.6y      0.0      0.0   0.0  cellular   

       Duration  Campaign  Pdays  Previous     Poutcome  Emp.var.rate  \
25903       299         2    999         0  NONEXISTENT          -0.1   

      Cons.price.idx Cons.conf.idx Euribor3m Nr.employed   Y            Date  \
25903           93,2           -42      4,12      5195,8  no  9-octubre-2015   

       Latitude  Longitude                                   Id_  
25903    31.634   -107.287  55b2d0d1-c5bb-41b6-ba2c-00057a35c2c2

In [66]:
#transformar valores 1 / 0 a yes, no
sp.transformar_booleanos(df_bankadditional)

Columnas booleanas detectadas: ['Default', 'Housing', 'Loan']

Muestra del DataFrame:


Age         Job   Marital            Education Default Housing Loan  \
11431  50.0    services  DIVORCED                  NaN      no      no   no   
1150   50.0  technician   MARRIED  professional.course      no      no   no   
16300  23.0      admin.    SINGLE  professional.course      no      no   no   
24472  31.0  technician    SINGLE    university.degree      no     yes  yes   
6438   51.0      admin.  DIVORCED    university.degree      no     yes   no   

         Contact  Duration  Campaign  Pdays  Previous     Poutcome  \
11431  telephone       250         3    999         0  NONEXISTENT   
1150   telephone       103         3    999         0  NONEXISTENT   
16300   cellular       187         1    999         0  NONEXISTENT   
24472   cellular       118         2    999         0  NONEXISTENT   
6438   telephone        72        18    999         0  NONEXISTENT   

       Emp.var.rate Cons.price.idx Cons.conf.idx Euribor3m Nr.employed   Y  \
11431           1.4         94,465         -41,8       NaN      5228,1  no   
1150            1.1         93,994         -36,4     4,855        5191  no   
16300           1.4         93,918         -42,7     4,963      5228,1  no   
24472          -0.1           93,2           -42       NaN      5195,8  no   
6438            1.1         93,994         -36,4     4,857        5191  no   

                     Date  Latitude  Longitude  \
11431        5-junio-2018    28.411   -120.315   
1150        26-enero-2018    33.307    -75.252   
16300    1-noviembre-2019    31.106    -99.251   
24472        7-marzo-2016    48.292    -87.030   
6438   19-septiembre-2016    24.536    -80.942   

                                        Id_  
11431  027cb068-bd5a-4dd9-ac15-8760fbc492d9  
1150   0863701e-cac1-41d5-b1d2-5555ab58ffc2  
16300  b8cd688f-e1e2-4a8d-8b39-ee1ad8909f62  
24472  0f3ad13e-aa88-46fe-8070-f5fa27996185  
6438   02d8bfbf-03cc-4dfd-976d-eeb546864baf

In [67]:
#cambiar , por . y convertir a float
sp.convertir_a_float(df_bankadditional)

Se han modificado las siguientes columnas:['Cons.price.idx', 'Cons.conf.idx', 'Euribor3m', 'Nr.employed']


Age          Job Marital Education Default Housing Loan   Contact  \
13967  26.0  blue-collar  SINGLE  basic.9y     NaN     yes   no  cellular   

       Duration  Campaign  Pdays  Previous     Poutcome  Emp.var.rate  \
13967       151         2    999         0  NONEXISTENT           1.4   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y  \
13967          93.918          -42.7      4.963       5228.1  no   

                    Date  Latitude  Longitude  \
13967  12-diciembre-2017    35.306   -101.423   

                                        Id_  
13967  688c247b-b6d8-4444-a018-da4ca9e9f073

In [68]:
#redondear decimal y convertir a int
sp.float_a_int(df_bankadditional,'Nr.employed')

La columna Nr.employed ha sido cambiada a numero entero y redondeada
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 23 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             37880 non-null  float64
 1   Job             42655 non-null  object 
 2   Marital         42915 non-null  object 
 3   Education       41193 non-null  object 
 4   Default         34019 non-null  object 
 5   Housing         41974 non-null  object 
 6   Loan            41974 non-null  object 
 7   Contact         43000 non-null  object 
 8   Duration        43000 non-null  int64  
 9   Campaign        43000 non-null  int64  
 10  Pdays           43000 non-null  int64  
 11  Previous        43000 non-null  int64  
 12  Poutcome        43000 non-null  object 
 13  Emp.var.rate    43000 non-null  float64
 14  Cons.price.idx  42529 non-null  float64
 15  Cons.conf.idx   43000 non-null  float64
 16  Euribor

None

Age         Job Marital            Education Default Housing Loan  \
21741  40.0  technician  SINGLE  professional.course      no     yes   no   

        Contact  Duration  Campaign  Pdays  Previous     Poutcome  \
21741  cellular       142         1    999         0  NONEXISTENT   

       Emp.var.rate  Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed  \
21741           1.4          93.444          -36.1        NaN         5228   

        Y          Date  Latitude  Longitude  \
21741  no  7-marzo-2017    43.089    -88.287   

                                        Id_  
21741  d810a4d4-fdc9-4bfd-ae53-b9fcdcfba631

In [69]:
sp.eliminar_columna(df_bankadditional,'Pdays')

Se ha eliminado la columna Pdays


Age         Job   Marital            Education Default Housing Loan  \
2792  32.0  technician  DIVORCED  professional.course      no     yes   no   

        Contact  Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
2792  telephone        70         2         0  NONEXISTENT           1.1   

      Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y  \
2792          93.994          -36.4      4.859         5191  no   

                 Date  Latitude  Longitude  \
2792  27-febrero-2018    40.225    -89.901   

                                       Id_  
2792  d7783d5b-528f-4a05-b93a-84fb4af29e8f

In [70]:
#convertir la columna a fecha
sp.convertir_fecha(df_bankadditional,'Date','mixed')

Se ha convertido la columna Date a fecha
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 22 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Age             37880 non-null  float64       
 1   Job             42655 non-null  object        
 2   Marital         42915 non-null  object        
 3   Education       41193 non-null  object        
 4   Default         34019 non-null  object        
 5   Housing         41974 non-null  object        
 6   Loan            41974 non-null  object        
 7   Contact         43000 non-null  object        
 8   Duration        43000 non-null  int64         
 9   Campaign        43000 non-null  int64         
 10  Previous        43000 non-null  int64         
 11  Poutcome        43000 non-null  object        
 12  Emp.var.rate    43000 non-null  float64       
 13  Cons.price.idx  42529 non-null  float64       
 14  Cons.conf.idx

Age          Job   Marital Education Default Housing Loan   Contact  \
35224  53.0  blue-collar  DIVORCED  basic.9y      no     yes   no  cellular   

       Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
35224       509         3         0  NONEXISTENT          -1.8   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  \
35224          92.893          -46.2       1.25         5099  no 2016-09-15   

       Latitude  Longitude                                   Id_  
35224    47.624    -84.932  36eda351-e8ff-4d76-b1a6-3b600c333bf8

In [71]:
#Convertir marital a minúsculas
sp.convertir_minusculas(df_bankadditional,'Marital')

Age         Job Marital          Education Default Housing Loan  \
22404  NaN  technician  single  university.degree      no     yes   no   

        Contact  Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
22404  cellular        55         3         0  NONEXISTENT           1.4   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  \
22404          93.444          -36.1        NaN         5228  no 2019-12-05   

       Latitude  Longitude                                   Id_  
22404    31.353   -107.896  50767884-a869-4c84-aa18-d5f9f52b3346

In [72]:
#convertir valores a minúsculas en la columna poutcome
sp.convertir_minusculas(df_bankadditional,'Poutcome')

Age      Job Marital    Education Default Housing Loan   Contact  \
40186  21.0  student  single  high.school      no     yes   no  cellular   

       Duration  Campaign  Previous Poutcome  Emp.var.rate  Cons.price.idx  \
40186       326         2         1  failure          -1.7          94.027   

       Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  Latitude  \
40186          -38.3      0.899         4992  no 2019-03-25     34.16   

       Longitude                                   Id_  
40186   -117.401  851c79b4-d0db-4b3f-9fb4-0384e4541ebc

In [73]:
#CAMBIAR . POR _
sp.punto_barrabaja(df_bankadditional,'Education')

Se ha cambiado '.' por '_' en la columna Education


Age          Job  Marital Education Default Housing Loan    Contact  \
10187  32.0  blue-collar  married  basic_9y      no     yes  yes  telephone   

       Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
10187       452         1         0  nonexistent           1.4   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  \
10187          94.465          -41.8       4.96         5228  no 2015-12-06   

       Latitude  Longitude                                   Id_  
10187    39.922    -70.529  a1ba00e4-ccae-4e63-9ddf-7485c5e14e92

In [74]:
#Cambiar - por _
sp.guion_barrabaja(df_bankadditional,'Job')

Se ha cambiado '-' por '_' en la columna Job


Job
admin.           10873
blue_collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self_employed     1489
housemaid         1123
unemployed        1063
student            903
nan                345
Name: count, dtype: int64

In [75]:
#Cambiar el nombre de la columna Id_ por ID
sp.cambiar_columna(df_bankadditional,'Id_','ID')

Se ha cambiado el nombre de la columna Id_ por ID


Age     Job  Marital    Education Default Housing Loan    Contact  \
664  38.0  admin.  married  high_school      no      no   no  telephone   

     Duration  Campaign  Previous     Poutcome  Emp.var.rate  Cons.price.idx  \
664       112         4         0  nonexistent           1.1          93.994   

     Cons.conf.idx  Euribor3m  Nr.employed   Y       Date  Latitude  \
664          -36.4      4.857         5191  no 2017-06-11    45.852   

     Longitude                                    ID  
664      -68.6  0f35f798-26b8-4da4-a056-92b8eb0d9a22

In [76]:
#Cambiar el nombre de la columna Y por Subscribed
sp.cambiar_columna(df_bankadditional,'Y','Subscribed')

Se ha cambiado el nombre de la columna Y por Subscribed


Age     Job   Marital          Education Default Housing Loan  \
19166  40.0  admin.  divorced  university_degree      no     yes  yes   

        Contact  Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
19166  cellular        65         1         0  nonexistent           1.4   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed Subscribed  \
19166          93.444          -36.1        NaN         5228         no   

            Date  Latitude  Longitude                                    ID  
19166 2015-04-27    32.481    -80.556  3686a147-c765-43ae-a210-650f40085903

In [77]:
df_bankadditional.columns

Index(['Age', 'Job', 'Marital', 'Education', 'Default', 'Housing', 'Loan',
       'Contact', 'Duration', 'Campaign', 'Previous', 'Poutcome',
       'Emp.var.rate', 'Cons.price.idx', 'Cons.conf.idx', 'Euribor3m',
       'Nr.employed', 'Subscribed', 'Date', 'Latitude', 'Longitude', 'ID'],
      dtype='object')

Se unen los dos dataframes, df_bank_limpio y df_customerdetails

In [78]:
df_bank_merge=df_bankadditional.merge(df_customerdetails,how='outer',on='ID' )

In [79]:
sp.eda_preliminar(df_bank_merge)

Age         Job   Marital          Education Default Housing Loan  \
36101   NaN    services  divorced        high_school      no      no  yes   
27015   NaN    services  divorced  university_degree      no     yes   no   
33534  56.0  management  divorced           basic_9y     NaN     yes   no   
25440  36.0      admin.    single  university_degree      no     yes   no   
8310    NaN   housemaid   married           basic_4y     NaN     yes   no   

         Contact  Duration  Campaign  Previous     Poutcome  Emp.var.rate  \
36101  telephone     176.0       2.0       0.0  nonexistent           1.1   
27015   cellular     335.0       1.0       1.0      failure          -1.8   
33534  telephone     313.0       2.0       0.0  nonexistent           1.1   
25440   cellular    1217.0       2.0       0.0  nonexistent           1.4   
8310    cellular     195.0       1.0       0.0  nonexistent           1.4   

       Cons.price.idx  Cons.conf.idx  Euribor3m  Nr.employed Subscribed  \
36101          93.994          -36.4        NaN       5191.0         no   
27015          92.893          -46.2      1.344       5099.0         no   
33534          93.994          -36.4        NaN       5191.0         no   
25440          93.444          -36.1      4.964       5228.0         no   
8310           93.444          -36.1      4.968       5228.0         no   

            Date  Latitude  Longitude                                    ID  \
36101 2015-02-11    25.930    -75.616  d64a45d7-d350-450d-a960-56d19cdbfaf9   
27015 2019-02-06    29.408    -96.732  a0008fb8-a8b4-4b64-9f2c-970328bbfcb7   
33534 2019-02-08    48.791   -105.395  c7076c75-9c01-4675-99f8-8f9427ea5503   
25440 2017-06-23    49.118   -108.368  970fdfc8-fccf-422d-aee9-a8a264666f63   
8310  2019-11-25    48.147    -86.633  31078b43-bc75-4654-b10c-30177342fbac   

       Income  Kidhome  Teenhome Dt_Customer  NumWebVisitsMonth  
36101  138943        2         1  2012-01-01                 13  
27015  121186        2         2  2014-08-09                 27  
33534   50509        2         0  2012-05-30                 27  
25440   41588        0         0  2013-02-20                  3  
8310    44347        0         2  2012-08-08                 12

--------------------
INFO
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 27 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Age                37880 non-null  float64       
 1   Job                43000 non-null  object        
 2   Marital            42915 non-null  object        
 3   Education          43000 non-null  object        
 4   Default            34019 non-null  object        
 5   Housing            41974 non-null  object        
 6   Loan               41974 non-null  object        
 7   Contact            43000 non-null  object        
 8   Duration           43000 non-null  float64       
 9   Campaign           43000 non-null  float64       
 10  Previous           43000 non-null  float64       
 11  Poutcome           43000 non-null  object        
 12  Emp.var.rate       43000 non-null  float64       
 13  Cons.price.idx     42529 non-null  

None

--------------------
NULOS


Age                  12.25
Job                   0.39
Marital               0.59
Education             0.39
Default              21.20
Housing               2.77
Loan                  2.77
Contact               0.39
Duration              0.39
Campaign              0.39
Previous              0.39
Poutcome              0.39
Emp.var.rate          0.39
Cons.price.idx        1.48
Cons.conf.idx         0.39
Euribor3m            21.83
Nr.employed           0.39
Subscribed            0.39
Date                  0.97
Latitude              0.39
Longitude             0.39
ID                    0.00
Income                0.00
Kidhome               0.00
Teenhome              0.00
Dt_Customer           0.00
NumWebVisitsMonth     0.00
dtype: float64

--------------------
duplicados
0
VALUE COUNT


Job
admin.           10873
blue_collar       9654
technician        7026
services          4162
management        3050
retired           1790
entrepreneur      1522
self_employed     1489
housemaid         1123
unemployed        1063
student            903
nan                345
Name: count, dtype: int64

--------------------


Marital
married     25999
single      12105
divorced     4811
Name: count, dtype: int64

--------------------


Education
university_degree      12722
high_school             9925
basic_9y                6309
professional_course     5477
basic_4y                4356
basic_6y                2386
nan                     1807
illiterate                18
Name: count, dtype: int64

--------------------


Default
no     34016
yes        3
Name: count, dtype: int64

--------------------


Housing
yes    22498
no     19476
Name: count, dtype: int64

--------------------


Loan
no     35442
yes     6532
Name: count, dtype: int64

--------------------


Contact
cellular     27396
telephone    15604
Name: count, dtype: int64

--------------------


Poutcome
nonexistent    37103
failure         4461
success         1436
Name: count, dtype: int64

--------------------


Subscribed
no     38156
yes     4844
Name: count, dtype: int64

--------------------


ID
fffe84df-4e63-4465-82de-0d9821d919bf    1
0000e811-006e-4404-b535-89bf6cd96553    1
ffdbe65e-3249-456c-a8f9-544935b40fdd    1
ffdbe151-4f82-4c19-b63c-841c91fc3084    1
ffda2b4b-0b68-493e-ac68-f81614c57f5d    1
                                       ..
0006c8fd-9dea-439d-a3b9-571b3901ff58    1
000625bd-cc08-4653-b8b7-1853c458527d    1
0004e1d1-958d-4abf-a57c-9b9c7be887a0    1
00024507-c59b-4eee-86d5-cc341b96eb6d    1
000165f9-20c0-4cb5-bd47-6233b92655c1    1
Name: count, Length: 43170, dtype: int64

--------------------
VALORES NUMÉRICOS


count                           mean  \
Age                37880.0                      39.977112   
Duration           43000.0                     257.739279   
Campaign           43000.0                       2.567233   
Previous           43000.0                       0.174023   
Emp.var.rate       43000.0                       0.077128   
Cons.price.idx     42529.0                      93.574219   
Cons.conf.idx      43000.0                     -40.509049   
Euribor3m          33744.0                       3.616521   
Nr.employed        43000.0                    5166.835721   
Date                 42752  2017-07-01 08:57:44.595808256   
Latitude           43000.0                      36.856697   
Longitude          43000.0                     -95.939067   
Income             43170.0                   93227.389622   
Kidhome            43170.0                       1.004749   
Teenhome           43170.0                       0.998633   
Dt_Customer          43170  2013-05-11 22:09:47.407922176   
NumWebVisitsMonth  43170.0                       16.59129   

                                   min                  25%  \
Age                               17.0                 32.0   
Duration                           0.0                102.0   
Campaign                           1.0                  1.0   
Previous                           0.0                  0.0   
Emp.var.rate                      -3.4                 -1.8   
Cons.price.idx                  92.201               93.075   
Cons.conf.idx                    -50.8                -42.7   
Euribor3m                        0.634                1.344   
Nr.employed                     4964.0               5099.0   
Date               2015-01-01 00:00:00  2016-04-04 00:00:00   
Latitude                        24.396             30.61475   
Longitude                     -124.997           -110.49425   
Income                          5841.0              49608.0   
Kidhome                            0.0                  0.0   
Teenhome                           0.0                  0.0   
Dt_Customer        2012-01-01 00:00:00  2012-07-16 00:00:00   
NumWebVisitsMonth                  1.0                  9.0   

                                   50%                  75%  \
Age                               38.0                 47.0   
Duration                         179.0                319.0   
Campaign                           2.0                  3.0   
Previous                           0.0                  0.0   
Emp.var.rate                       1.1                  1.4   
Cons.price.idx                  93.749               93.994   
Cons.conf.idx                    -41.8                -36.4   
Euribor3m                        4.857                4.961   
Nr.employed                     5191.0               5228.0   
Date               2017-07-04 00:00:00  2018-09-30 00:00:00   
Latitude                        36.761             43.11325   
Longitude                     -95.8995            -81.42775   
Income                         93009.5             136740.5   
Kidhome                            1.0                  2.0   
Teenhome                           1.0                  2.0   
Dt_Customer        2013-03-01 00:00:00  2014-03-28 00:00:00   
NumWebVisitsMonth                 17.0                 25.0   

                                   max           std  
Age                               98.0     10.437957  
Duration                        4918.0    258.666033  
Campaign                          56.0      2.772294  
Previous                           7.0      0.497366  
Emp.var.rate                       1.4      1.573898  
Cons.price.idx                  94.767      0.579548  
Cons.conf.idx                    -26.9      4.637186  
Euribor3m                        5.045      1.737117  
Nr.employed                     5228.0     72.337206  
Date               2019-12-31 00:00:00           NaN  
Latitude                        49.384      7.2

In [80]:
df_bank_merge.to_csv('bank_merge.csv')